### ISSUE TITLE
Implement reference Unstructured store API to upload TaskRun logs GCS store

### ISSUE THREAD

**Issue by tejal29** [tektoncd/pipeline#107]
## Expected Behavior
The Pipeline `TaskRun` logs should be uploaded to an endpoint and available to download later.
In our initial reference implementation we should support uploading to GCS. In the long run we should support other kinds of stores, and provide a default that does not require GCS.

## Actual Behavior
~As of #167 the logs will be streamed to a PVC. This volume will continue to exist after the TaskRun has completed. Once this task is done, that PVC should no longer be needed.~ (This functionality was removed in https://github.com/tektoncd/pipeline/pull/443)

Since we moved from init containers to containers for steps in #564 logs are available via the pod logs through kube, however there are still only limited guarantees about how long the logs will be available for.

## Steps to Reproduce the Problem

  1. Create a Task
  2. Create a TaskRun
  3. Wait for TaskRun to complete
  4.  Downloaded the uploaded results.

**Comment by bobcatfish** [comment]
@tejal29 this is a stretch goal for the milestone so I'm going to remove it from the required milestone tasks

**Comment by tejal29** [comment]
After feedback with Build working group, we decided to go with following approach.
The API will provide a Sink interface.
A Sink Service Definition will contain
- Endpoint : This is url where the sink is hosted.
- Path to upload to: for now we will be only focusing on uploading logs. 
  * Note: We will add support for uploading the Status for TaskRun or PipelineRun  in subsequent iteration. 
When we want to upload sat task runs logs, the reconciler will call the service which is running at `Endpoint` and upload all the logs at path prefix <Path>
<Endpoint>:<Path>/<taskruns><task run id>

This will be a Https Service running in our cluster or any other cluster which your cluster can access.

The Sink `Base` definition will look like this.
```
package sink
type URL string
type Base struct {
   Endpoint URL
   Path String
}
```
The reason, we have `Path` same service can be used to upload logs for multiple projects. They will reside in different Paths and not overlap.

e.g.:  I have implemented a GCSSink and installed in my cluster. It is running at "104.198.205.71:8080".
The GCS sink defintion will look this
```
package sink

type GCS struct {
  *sink.Base
   project string //GCP project.
}
```
I can define two GCS Sinks which point to 2 buckets "cluster1", "cluster2"
```
sink1 := sink.GCS {
  Endpoint: "104.198.205.71:8080",
  Path: "cluster1",
  Project: "test1",  // only test1 has write access to gcs bucket gs://cluster1
}

sink2 := sink.GCS {
  Endpoint: "104.198.205.71:8080",
  Path: "cluster2",
 Project: "test2",  // Similarly, only test2 has write access to gcs bucket gs://cluster2
}
```

Along with that,  Sink Interface need to handle 4 url requests "upload/taskruns/", "download/taskruns/" , "upload/pipelineruns/",  "download/taskruns/" 
(We can also support, Upload and Get with run type. Its up to the implementer ) 
Note: We can support partial upload and download of logs later.

The Task Reconciler will now make a HTTP request to this `sink.Endpoint/sink.Path/upload/taskruns/<id=x>,<contentstream=>`

The Design Question over here is how to Define a Sink for a pipeline or a Task run.
Should sink be defined cluster wide Custom Resource?
- this would mean, we will create a new custom resource for Sink. 
- Admins can create multiple Sinks.
- The Reconciler will fetch all sinks installed in your cluster and then upload logs to all sinks.


Should Sink be defined per Pipeline or Task?
- this would Sink will added to `PipelineParams`
- this would mean, now task and task run definition will list the sink it would want to sink the results too


/cc @ImJasonH and @bobcatfish  and @aaron-prindle  does this all make sense?

**Comment by sbwsg** [comment]
I'm closing this issue out as we have now circulated a [design doc](https://docs.google.com/document/d/1YJrRXY7S1VIzWrJUoY4Xy9NJzcx6fwaq3Cv41Gqhgs4) for logging in Tekton and the utility of information retained in this issue is limited due to its age.

I've opened #1155 to encompass the work of validating and implementing the proposed design and encourage anyone looking to get involved on this topic to add commentary, use cases and counterpoints to the design doc or github issue linked above. Cheers!
